relayer/src/event/monitor.rs

```diff
@@ -336,32 +354,55 @@ impl EventMonitor {
 
             let result = rt.block_on(async {
                 tokio::select! {
-                    Some(batch) = batches.next() => Ok(batch),
+                    Some(batch) = batches.next() => batch,
                     Some(err) = self.rx_err.recv() => Err(Error::WebSocketDriver(err)),
                 }
             });
 
             match result {
                 Ok(batch) => self.process_batch(batch).unwrap_or_else(|e| {
-                    warn!(chain.id = %self.chain_id, "{}", e);
+                    error!(chain.id = %self.chain_id, "{}", e);
                 }),
+                Err(Error::SubscriptionCancelled(reason)) => {
+                    error!(chain.id = %self.chain_id, "subscription cancelled, reason: {}", reason);
+
+                    self.propagate_error(Error::SubscriptionCancelled(reason))
+                        .unwrap_or_else(|e| {
+                            error!(chain.id = %self.chain_id, "{}", e);
+                        });
+
+                    // Reconnect to the WebSocket endpoint, and subscribe again to the queries.
+                    self.reconnect();
+
+                    // Abort this event loop, the `run` method will start a new one.
+                    // We can't just write `return self.run()` here because Rust
+                    // does not perform tail call optimization, and we would
+                    // thus potentially blow up the stack after many restarts.
+                    return Next::Continue;
+                }
                 Err(e) => {
                     error!(chain.id = %self.chain_id, "failed to collect events: {}", e);
 
-                    // Restart the event monitor, reconnect to the WebSocket endpoint,
-                    // and subscribe again to the queries.
-                    self.restart();
+                    // Reconnect to the WebSocket endpoint, and subscribe again to the queries.
+                    self.reconnect();
 
                     // Abort this event loop, the `run` method will start a new one.
                     // We can't just write `return self.run()` here because Rust
                     // does not perform tail call optimization, and we would
                     // thus potentially blow up the stack after many restarts.
-                    break;
+                    return Next::Continue;
                 }
             }
         }
+    }
 
-        Next::Continue
+    /// Propagate error to subscribers
```

Comment:
Could we expand on this comment to give some insight as to _who_ are the subscribers?

This code can be difficult to follow because of multiple layers of channels `monitor` <> `runtime` <> `supervisor` <> `workers`. (Not sure how accurate that is, btw.) It would improve understandability (and onboarding others) if we are more concrete here and specify the exact location where this error propagates to.

Reply:
Is the expanded comment more satisfactory?